File: backend/TradeX.Core/Models/BacktestTask.cs
using TradeX.Core.Abstractions;
using TradeX.Core.Enums;
using TradeX.Core.Events;

namespace TradeX.Core.Models;

public enum BacktestTaskStatus
{
    Pending,
    Running,
    Completed,
    Failed,
    Cancelled
}

public enum BacktestPhase
{
    Queued,
    FetchingData,
    Running
}

public class BacktestTask : AggregateRoot
{
    // EF Core 无参构造函数
    public BacktestTask() { }

    /// <summary>工厂方法：创建回测任务。</summary>
    public static BacktestTask Create(
        Guid strategyId, Guid exchangeId, string strategyName,
        string pair, string timeframe, decimal initialCapital,
        DateTime startAt, DateTime endAt, Guid createdBy,
        decimal? positionSize = null)
    {
        return new BacktestTask
        {
            StrategyId = strategyId,
            ExchangeId = exchangeId,
            StrategyName = strategyName,
            Pair = pair,
            Timeframe = timeframe,
            InitialCapital = initialCapital,
            PositionSize = positionSize,
            StartAt = startAt,
            EndAt = endAt,
            CreatedBy = createdBy
        };
    }

    public Guid Id { get; init; } = Guid.NewGuid();
    public Guid StrategyId { get; init; }
    public Guid ExchangeId { get; init; }
    public string StrategyName { get; set; } = string.Empty;
    public string Pair { get; set; } = string.Empty;
    public string Timeframe { get; set; } = string.Empty;
    public decimal InitialCapital { get; set; } = 1000m;
    public decimal? PositionSize { get; set; }
    public BacktestTaskStatus Status { get; set; } = BacktestTaskStatus.Pending;
    public BacktestPhase? Phase { get; set; }
    public DateTime StartAt { get; init; }
    public DateTime EndAt { get; init; }
    public Guid CreatedBy { get; init; }
    public DateTime CreatedAt { get; init; } = DateTime.UtcNow;
    public DateTime? CompletedAt { get; set; }

    // ─────────────── 领域方法 ───────────────

    public void Start()
    {
        if (Status != BacktestTaskStatus.Pending)
            throw new InvalidOperationException($"回测任务 {Id} 状态为 {Status}，不能启动");
        Status = BacktestTaskStatus.Running;
        Phase = BacktestPhase.FetchingData;
        AddDomainEvent(new BacktestStartedEvent(Id));
    }

    public void Complete(decimal? finalValue = null, decimal? totalReturnPercent = null)
    {
        if (Status != BacktestTaskStatus.Running)
            throw new InvalidOperationException($"回测任务 {Id} 状态为 {Status}，不能标记完成");
        Status = BacktestTaskStatus.Completed;
        Phase = null;
        CompletedAt = DateTime.UtcNow;
        AddDomainEvent(new BacktestCompletedEvent(Id, finalValue ?? 0, totalReturnPercent ?? 0));
    }

    public void Fail(string? reason = null)
    {
        Status = BacktestTaskStatus.Failed;
        Phase = null;
        CompletedAt = DateTime.UtcNow;
    }

    public void Cancel()
    {
        if (Status == BacktestTaskStatus.Completed || Status == BacktestTaskStatus.Failed)
            return;
        Status = BacktestTaskStatus.Cancelled;
        Phase = null;
        CompletedAt = DateTime.UtcNow;
    }
}

File: backend-go/internal/infra/ent/schema/backtest_task.go
package schema

import (
	"time"

	"entgo.io/ent"
	"entgo.io/ent/dialect"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/index"
	"github.com/google/uuid"
)

type BacktestTask struct {
	ent.Schema
}

func (BacktestTask) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("id", uuid.UUID{}).Default(uuid.New),
		field.UUID("strategy_id", uuid.UUID{}),
		field.String("strategy_name").MaxLen(200).Default(""),
		field.UUID("created_by", uuid.UUID{}).Default(uuid.New),
		field.UUID("exchange_id", uuid.UUID{}),
		field.String("pair").MaxLen(50),
		field.String("timeframe").MaxLen(10),
		field.Float("initial_capital").
			SchemaType(map[string]string{dialect.Postgres: "numeric(20,8)"}),
		field.Float("position_size").
			Optional().
			Nillable().
			SchemaType(map[string]string{dialect.Postgres: "numeric(20,8)"}),
		field.Time("start_at"),
		field.Time("end_at"),
		field.Time("completed_at").Optional().Nillable(),
		field.Enum("status").
			Values("Pending", "Running", "Completed", "Failed", "Cancelled").
			Default("Pending"),
		field.Enum("phase").
			Values("Queued", "FetchingData", "Running").
			Optional(),
		field.Time("created_at").Default(time.Now).Immutable(),
	}
}

func (BacktestTask) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("result", BacktestResult.Type).
			Unique().
			StorageKey(edge.Column("result_id")),
	}
}

func (BacktestTask) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("status"),
		index.Fields("strategy_id"),
		index.Fields("exchange_id"),
	}
}
